### PR TITLE
feat(permission): add facade api

### LINKED ISSUE #2730: [Feature] Provide a partially uneditable facade API
### 初始清单

- [X] 这真的是个问题吗？
- [X] 我已经在 Github Issues 中搜索过了，但没有找到类似的问题。

### 问题

存在很多不区分权限只有部分单元格能编辑的场景，现在只能通过WorksheetEditPermission让整个工作表不能编辑或者通过保护区域控制（保护区域只能控制其他用户不能编辑）

### 解决方案

类似setEditable一样，提供一个局部不可编辑的facade api

### PR DESCRIPTION
**Pull request by ybzky**
close #2730 

## Pull Request Checklist

- [ ] Related tickets or issues have been linked in the PR description (or missing issue).
- [ ] [Naming convention](https://github.com/dream-num/univer/blob/dev/docs/NAMING_CONVENTION.md) is followed (**do please** check it especially when you created new plugins, commands and resources).
- [ ] Unit tests have been added for the changes (if applicable).
- [ ] Breaking changes have been documented (or no breaking changes introduced in this PR).

### CODE REVIEW

**Comment by hexf00** [comment]
与 https://developers.google.com/apps-script/reference/spreadsheet/protection?hl=zh-cn api 不一样?

**Comment by ybzky** [comment]
> 与 https://developers.google.com/apps-script/reference/spreadsheet/protection?hl=zh-cn api 不一样?

不一样 也不太好写的一样，google的更多的倾向于user相关的操作，我们这个提供的是范围和不同的功能点位的增删改逻辑

**Comment by hexf00** [comment]
> > 与 https://developers.google.com/apps-script/reference/spreadsheet/protection?hl=zh-cn api 不一样?
> 
> 不一样 也不太好写的一样，google的更多的倾向于user相关的操作，我们这个提供的是范围和不同的功能点位的增删改逻辑

了解了，能补下测试用例吗，在测试用例中演示下使用例子

File: packages/facade/src/apis/sheets/f-permission.ts [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+    /**
+     * This function is used to add a base permission for a worksheet.
+     *
+     * @param {string} unitId - The unique identifier of the workbook for which the permission is being set.
+     * @param {string} subUnitId - The unique identifier of the worksheet for which the permission is being set.
+     *
+     * @returns {string | undefined} - The permission id of the added permission, or undefined if the permission was add failed.
+     */
+    addWorksheetBasePermission(unitId: string, subUnitId: string) {
+        const permissionId = `permissionId_${generateRandomId(8)}`;
```
Comment (Gggpound) [inline packages/facade/src/apis/sheets/f-permission.ts:77]: 有对应的service去生成。

**Comment by Gggpound** [comment]
https://developers.google.cn/apps-script/reference/spreadsheet/protection?hl=zh-cn#methods 
我这边的输入是尽量对齐google的api

File: packages/facade/src/apis/sheets/f-permission.ts (hunk 1 above)
Comment (ybzky) [reply packages/facade/src/apis/sheets/f-permission.ts]: 已修改，但是目前只传了type进去，没有协同人员以及scope相关的参数

**Review by hexf00 — APPROVED** [review]

File: packages/facade/src/apis/sheets/f-permission.ts [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+    }
+
+    /**
+     * Removes the range protection from the worksheet.
+     *
+     * @param {string} unitId - The unique identifier of the workbook.
+     * @param {string} subUnitId - The unique identifier of the worksheet.
+     * @param {string[]} ruleIds - The rule IDs of the range protection to be removed.
+     */
+    removeRangeProtection(unitId: string, subUnitId: string, ruleIds: string[]): void {
```
[the remark this answers is not in the document]
Comment (ybzky) [reply packages/facade/src/apis/sheets/f-permission.ts:204]: added

File: packages/facade/src/apis/sheets/f-permission.ts [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+     * This method finds the rule by unitId, subUnitId, and ruleId, and updates the rule with the provided ranges.
+     * It checks for overlaps with existing ranges in the same subunit and shows an error message if any overlap is detected.
+     * If no overlap is found, it executes the command to update the range protection with the new ranges.
+     *
+     * @param {string} unitId - The unique identifier of the workbook.
+     * @param {string} subUnitId - The unique identifier of the worksheet within the workbook.
+     * @param {string} ruleId - The ruleId of the range protection rule that is being updated.
+     * @param {IRange[]} ranges - The array of new ranges to be set for the range protection rule.
+     */
+    setRangeProtectionRanges(unitId: string, subUnitId: string, ruleId: string, ranges: IRange[]): void {
```
[the remark this answers is not in the document]
Comment (ybzky) [reply packages/facade/src/apis/sheets/f-permission.ts:254]: added

**Pull request merged**